Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionSidebar.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+
+#import "WebExtensionContext.h"
+#import "WebExtensionTab.h"
+#import "WebExtensionWindow.h"
+
+namespace WebKit {
+
+static NSString * const sidebarActionManifestKey = @"sidebar_action";
+static NSString * const sidePanelManifestKey = @"side_panel";
+static NSString * const sidebarActionTitleKey = @"default_title";
+static NSString * const sidebarActionPathKey = @"default_panel";
+static NSString * const sidePanelPathKey = @"default_path";
+
+static NSString * const fallbackPath = @"about:blank";
+static NSString * const fallbackTitle = @"";
+
+static std::optional<RetainPtr<NSDictionary<NSString *, id>>> ensureDict(id maybeDict)
+{
+    if (!maybeDict)
+        return std::nullopt;
+    if (![maybeDict isKindOfClass:NSDictionary.class])
+        return std::nullopt;
+    return (NSDictionary<NSString *, id> *) maybeDict;
+}
```

Comment:
Could move to `CocoaHelpers` and be a generic template. Something like:

```suggestion
template<typename T>
std::optional<RetainPtr<T>> optional_objc_cast(id object)
{
    if (auto *result = dynamic_objc_cast<T>(object))
        return result;
    return std::nullopt;
}
```

Reply:
I ended up removing this in favor of existing nullable pointer methods for the parsing code I'm moving to `WebExtension` -- it feels like it fits in better with the other existing `WebExtension` methods than something based on `std::optional` would, and I can just perform the conversion to optional in `WebExtensionSidebarCocoa.cpp`.